include/ginkgo/core/base/matrix_assembly_data.hpp

```diff
@@ -0,0 +1,224 @@
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+
+
+namespace gko {
+
+
+namespace detail {
+
+
+/**
+ * Type used to store indices used as keys in the unordered_map storing
+ * the nonzeros.
+ */
+template <typename IndexType>
+struct symbolic_nonzero {
+    using index_type = IndexType;
+    symbolic_nonzero() = default;
+
+    symbolic_nonzero(index_type r, index_type c) noexcept : row(r), column(c) {}
+
+    bool operator==(symbolic_nonzero other) const noexcept
+    {
+        return std::tie(this->row, this->column) ==
+               std::tie(other.row, other.column);
+    }
+
+    bool operator!=(symbolic_nonzero other) const noexcept
+    {
+        return !(*this == other);
+    }
+
+    index_type row;
+    index_type column;
+};
+
+
+template <typename IndexType>
+struct symbolic_nonzero_hash {
+    explicit symbolic_nonzero_hash(size_type num_cols) noexcept
+        : num_cols_{num_cols}
+    {}
+
+    std::size_t operator()(symbolic_nonzero<IndexType> nnz) const noexcept
+    {
+        return nnz.row * num_cols_ + nnz.column;
+    }
+
+    std::size_t num_cols_;
+};
+
+
+}  // namespace detail
+
+
+/**
+ * This structure is used as an intermediate type to assemble a sparse matrix.
+ *
+ * The matrix is stored as a set of nonzero elements, where each element is
+ * a triple of the form (row_index, column_index, value).
+ *
+ * New values can be added by using the matrix_assembly_data::add_entry or
+ * matrix_assembly_data::add_entries functions.
+ *
+ * @tparam ValueType  type of matrix values stored in the structure
+ * @tparam IndexType  type of matrix indexes stored in the structure
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class matrix_assembly_data {
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    explicit matrix_assembly_data(dim<2> size)
+        : size_{size},
+          nonzeros_(0, detail::symbolic_nonzero_hash<index_type>(size_[1]))
+    {}
+
+    /**
+     * Sets the matrix value at (row, col).
+     * If there is an existing value, it will be replaced by the sum of the
+     * existing and new value otherwise it will be.
+     */
+    void add_value(index_type row, index_type col, value_type val)
+    {
+        auto ind = detail::symbolic_nonzero<index_type>{row, col};
+        nonzeros_[ind] += val;
+    }
+
+    /**
+     * Sets the matrix value at (row, col).
+     * If there is an existing value, it will be overwritten by the new value.
+     */
+    void set_value(index_type row, index_type col, value_type val)
+    {
+        auto ind = detail::symbolic_nonzero<index_type>{row, col};
+        nonzeros_[ind] = val;
+    }
+
+    /**
+     * Gets the matrix value at (row, col).
+     * Returns zero if it doesn't exist.
+     */
+    value_type get_value(index_type row, index_type col)
+    {
+        auto ind = detail::symbolic_nonzero<index_type>{row, col};
+        auto it = nonzeros_.find(ind);
```

Comment:
nit: both can be const here:
```suggestion
        const auto ind = detail::symbolic_nonzero<index_type>{row, col};
        const auto it = nonzeros_.find(ind);
```